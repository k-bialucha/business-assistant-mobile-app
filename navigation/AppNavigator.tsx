import React from 'react';
import { createBottomTabNavigator } from '@react-navigation/bottom-tabs';
import { createStackNavigator } from '@react-navigation/stack';

import DashboardScreen from '../screens/DashboardScreen';
import SalesScreen from '../screens/SalesScreen';
import CostsScreen from '../screens/CostsScreen';
import CostsEntryScreen from '../screens/CostsEntryScreen';
import SalesEntryScreen from '../screens/SalesEntryScreen';

const DashboardStackNavigator = createStackNavigator();

const DashboardNavigator = () => {
  return (
    <DashboardStackNavigator.Navigator>
      <DashboardStackNavigator.Screen
        name="Dashboard"
        component={DashboardScreen}
      />
    </DashboardStackNavigator.Navigator>
  );
};

const SalesStackNavigator = createStackNavigator();

const SalesNavigator = () => {
  return (
    <SalesStackNavigator.Navigator>
      <SalesStackNavigator.Screen name="SalesList" component={SalesScreen} />
      <SalesStackNavigator.Screen
        name="SalesEntry"
        component={SalesEntryScreen}
      />
    </SalesStackNavigator.Navigator>
  );
};

const CostsStackNavigator = createStackNavigator();

const CostsNavigator = () => {
  return (
    <CostsStackNavigator.Navigator>
      <CostsStackNavigator.Screen name="CostsList" component={CostsScreen} />
      <CostsStackNavigator.Screen
        name="CostsEntry"
        component={CostsEntryScreen}
      />
    </CostsStackNavigator.Navigator>
  );
};

const AppTabNavigator = createBottomTabNavigator();

const AppNavigator = () => {
  return (
    <AppTabNavigator.Navigator>
      <AppTabNavigator.Screen name="Dashboard" component={DashboardNavigator} />
      <AppTabNavigator.Screen name="Sales" component={SalesNavigator} />
      <AppTabNavigator.Screen name="Costs" component={CostsNavigator} />
    </AppTabNavigator.Navigator>
  );
};

export default AppNavigator;
