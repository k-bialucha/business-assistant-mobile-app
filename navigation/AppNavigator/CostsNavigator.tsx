import React from 'react';

import { createStackNavigator } from '@react-navigation/stack';

import CostsEntryScreen from '../../screens/CostsEntryScreen';
import CostsScreen, { CostsScreenNavOptions } from '../../screens/CostsScreen';

import defaultScreenOptions from './defaultScreenOptions';
import { ScreenProps } from './index';

const CostsStackNavigator = createStackNavigator();

type Props = ScreenProps<'Costs'>;

const CostsNavigator: React.FC<Props> = () => {
  return (
    <CostsStackNavigator.Navigator screenOptions={defaultScreenOptions}>
      <CostsStackNavigator.Screen
        name="CostsList"
        component={CostsScreen}
        options={CostsScreenNavOptions}
      />
      <CostsStackNavigator.Screen
        name="CostsEntry"
        component={CostsEntryScreen}
      />
    </CostsStackNavigator.Navigator>
  );
};

export default CostsNavigator;
