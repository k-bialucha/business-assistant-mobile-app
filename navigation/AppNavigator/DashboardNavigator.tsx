import React from 'react';
import {
  createStackNavigator,
  StackNavigationProp,
} from '@react-navigation/stack';
import { RouteProp } from '@react-navigation/native';

import { RootStackParamList } from '.';
import defaultScreenOptions from './defaultScreenOptions';

import DashboardScreen from '../../screens/DashboardScreen';

const DashboardStackNavigator = createStackNavigator<RootStackParamList>();

type ScreenNavigationProp = StackNavigationProp<
  RootStackParamList,
  'Dashboard'
>;
type ScreenRouteProp = RouteProp<RootStackParamList, 'Dashboard'>;

type DashboardNavigationProps = React.FC<{
  navigation: ScreenNavigationProp;
  route: ScreenRouteProp;
}>;

const DashboardNavigator: DashboardNavigationProps = () => {
  return (
    <DashboardStackNavigator.Navigator screenOptions={defaultScreenOptions}>
      <DashboardStackNavigator.Screen
        name="Dashboard"
        component={DashboardScreen}
      />
    </DashboardStackNavigator.Navigator>
  );
};

export default DashboardNavigator;
