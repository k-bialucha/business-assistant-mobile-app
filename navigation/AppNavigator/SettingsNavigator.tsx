import React from 'react';

import { CompositeNavigationProp, RouteProp } from '@react-navigation/native';
import {
  createStackNavigator,
  StackNavigationProp,
} from '@react-navigation/stack';

import AuthScreen from '../../screens/AuthScreen';

import defaultScreenOptions from './defaultScreenOptions';
import {
  AppNavigatorNavigationProp,
  NavigationData as AppNavigatorNavigationData,
} from './index';

type ParamList = {
  Settings: undefined;
};

const SettingsStackNavigator = createStackNavigator<ParamList>();

type Props = AppNavigatorNavigationData<'Settings'>;

const SettingsNavigator: React.FC<Props> = () => {
  return (
    <SettingsStackNavigator.Navigator screenOptions={defaultScreenOptions}>
      <SettingsStackNavigator.Screen name="Settings" component={AuthScreen} />
    </SettingsStackNavigator.Navigator>
  );
};

export default SettingsNavigator;

// combine navigation prop for nested navigators
// (AppNavigator data + this StackNavigator data)
export type NavigationPropCombined<
  RouteName extends keyof ParamList
> = CompositeNavigationProp<
  AppNavigatorNavigationProp<'Settings'>,
  StackNavigationProp<ParamList, RouteName>
>;

// export type so screens can get typing
export type NavigationData<RouteName extends keyof ParamList> = {
  navigation: NavigationPropCombined<RouteName>;
  route: RouteProp<ParamList, RouteName>;
};
