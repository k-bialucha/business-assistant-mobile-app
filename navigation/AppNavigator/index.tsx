import React from 'react';
import { Platform } from 'react-native';

import { Ionicons } from '@expo/vector-icons';
import { createBottomTabNavigator } from '@react-navigation/bottom-tabs';
import { RouteProp } from '@react-navigation/native';
import { StackNavigationProp } from '@react-navigation/stack';

import Colors from '../../theme/Colors';

import CostsNavigator from './CostsNavigator';
import DashboardNavigator from './DashboardNavigator';
import SalesNavigator from './SalesNavigator';
import SettingsNavigator from './SettingsNavigator';

export type RootStackParamList = {
  Dashboard: undefined;
  Sales: undefined;
  Costs: undefined;
  Settings: undefined;
};

export type ScreenProps<RouteName extends keyof RootStackParamList> = {
  navigation: StackNavigationProp<RootStackParamList, RouteName>;
  route: RouteProp<RootStackParamList, RouteName>;
};

const AppTabNavigator = createBottomTabNavigator<RootStackParamList>();

const AppNavigator: React.FC<{}> = () => {
  return (
    <AppTabNavigator.Navigator
      initialRouteName="Dashboard"
      screenOptions={({ route }) => ({
        tabBarIcon: ({ color, size }) => {
          const icons = {
            Dashboard: Platform.OS === 'android' ? 'md-stats' : 'ios-stats',
            Sales:
              Platform.OS === 'android' ? 'md-trending-up' : 'ios-trending-up',
            Costs:
              Platform.OS === 'android'
                ? 'md-trending-down'
                : 'ios-trending-down',
            Settings:
              Platform.OS === 'android' ? 'md-settings' : 'ios-settings',
          };

          return (
            <Ionicons name={icons[route.name]} size={size} color={color} />
          );
        },
      })}
      tabBarOptions={{
        showLabel: false,
        activeTintColor: Colors.navyBlue,
        inactiveTintColor: Colors.silver,
      }}
    >
      <AppTabNavigator.Screen name="Sales" component={SalesNavigator} />
      <AppTabNavigator.Screen name="Dashboard" component={DashboardNavigator} />
      <AppTabNavigator.Screen name="Costs" component={CostsNavigator} />
      <AppTabNavigator.Screen name="Settings" component={SettingsNavigator} />
    </AppTabNavigator.Navigator>
  );
};

export default AppNavigator;
