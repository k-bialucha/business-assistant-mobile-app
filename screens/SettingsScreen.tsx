import React from 'react';
import { Button } from 'react-native';

import { useDispatch } from 'react-redux';

import { NavigationData } from '../navigation/AppNavigator/SettingsNavigator';
import { logout } from '../redux/auth';

import { StyledView } from './SettingsScreen.styled';

type Props = NavigationData<'Settings'>;

const SettingsScreen: React.FC<Props> = () => {
  const dispatch = useDispatch();

  return (
    <StyledView>
      {/* TODO: show user name/email and picture if existing */}
      <Button
        title="Log out"
        onPress={() => {
          dispatch(logout());
        }}
      />
    </StyledView>
  );
};

export default SettingsScreen;
