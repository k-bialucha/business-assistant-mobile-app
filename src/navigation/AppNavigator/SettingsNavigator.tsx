import React from 'react';

import { CompositeNavigationProp, RouteProp } from '@react-navigation/native';
import {
  createStackNavigator,
  StackNavigationProp,
} from '@react-navigation/stack';

import SettingsScreen from '~/screens/SettingsScreen';
import UserSettingsScreen from '~/screens/UserSettingsScreen';
import { UserSettingsScreenNavOptions } from '~/screens/UserSettingsScreen/UserSettingsScreen';

import defaultScreenOptions from './defaultScreenOptions';
import {
  AppNavigatorNavigationProp,
  NavigationData as AppNavigatorNavigationData,
} from './index';

type ParamList = {
  Settings: undefined;
  UserSettings: undefined;
};

const SettingsStackNavigator = createStackNavigator<ParamList>();

type Props = AppNavigatorNavigationData<'Settings'>;

const SettingsNavigator: React.FC<Props> = () => {
  return (
    <SettingsStackNavigator.Navigator screenOptions={defaultScreenOptions}>
      <SettingsStackNavigator.Screen
        name="Settings"
        component={SettingsScreen}
      />
      <SettingsStackNavigator.Screen
        name="UserSettings"
        component={UserSettingsScreen}
        options={UserSettingsScreenNavOptions}
      />
    </SettingsStackNavigator.Navigator>
  );
};

export default SettingsNavigator;

// combine navigation prop for nested navigators
// (AppNavigator data + this StackNavigator data)
export type NavigationPropCombined<
  RouteName extends keyof ParamList
> = CompositeNavigationProp<
  AppNavigatorNavigationProp<'Settings'>,
  StackNavigationProp<ParamList, RouteName>
>;

// export type so screens can get typing
export type NavigationData<RouteName extends keyof ParamList> = {
  navigation: NavigationPropCombined<RouteName>;
  route: RouteProp<ParamList, RouteName>;
};
