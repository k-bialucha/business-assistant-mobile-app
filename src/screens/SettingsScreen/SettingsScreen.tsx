import React from 'react';
import { Alert, TouchableWithoutFeedback } from 'react-native';

import { ListItem } from 'react-native-elements';
import { useDispatch, useSelector } from 'react-redux';

import { LanguageSwitcher } from '~/components/LanguageSwitcher';
import { useAppTranslation } from '~/hooks/useAppTranslation';
import { NavigationData } from '~/navigation/AppNavigator/SettingsNavigator';
import { logout } from '~/redux/auth';
import { RootState } from '~/redux/rootReducer';
import Colors from '~/theme/Colors';

import { StyledView } from './SettingsScreen.styled';

type Props = NavigationData<'Settings'>;

const SettingsScreen: React.FC<Props> = ({ navigation }) => {
  const dispatch = useDispatch();
  const { username, userImage } = useSelector((state: RootState) => state.auth);
  const { t } = useAppTranslation();

  if (!username) return null;

  return (
    <StyledView>
      <TouchableWithoutFeedback
        onPress={() => navigation.navigate('UserSettings')}
      >
        <ListItem
          leftAvatar={{
            rounded: true,
            ...(userImage ? { source: { uri: userImage } } : {}),
          }}
          title={username}
          titleStyle={{
            color: Colors.navyBlue,
            fontSize: 20,
          }}
          style={{ marginTop: 35, marginBottom: 35 }}
          chevron
        />
      </TouchableWithoutFeedback>
      <TouchableWithoutFeedback
        onPress={() => Alert.alert('"Open view screen with some settings"')}
      >
        <ListItem
          title="Test Option 1"
          titleStyle={{
            color: Colors.navyBlue,
          }}
          chevron
          style={{ marginBottom: 5 }}
        />
      </TouchableWithoutFeedback>
      <TouchableWithoutFeedback
        onPress={() => Alert.alert('"Open view screen with some settings"')}
      >
        <ListItem
          title="Test Option 2"
          titleStyle={{
            color: Colors.navyBlue,
          }}
          style={{ marginBottom: 5 }}
          chevron
        />
      </TouchableWithoutFeedback>
      <TouchableWithoutFeedback
        onPress={() => {
          dispatch(logout());
        }}
      >
        <ListItem
          title={t('Logout')}
          titleStyle={{
            color: Colors.errorRed,
          }}
          style={{ marginTop: 30 }}
          contentContainerStyle={{ alignItems: 'center' }}
        />
      </TouchableWithoutFeedback>
      <LanguageSwitcher />
    </StyledView>
  );
};

export default SettingsScreen;
