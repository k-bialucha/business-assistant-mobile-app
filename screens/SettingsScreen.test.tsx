import React from 'react';

import { shallow } from 'enzyme';

import { NavigationData } from '../navigation/AppNavigator/SettingsNavigator';

import SettingsScreen from './SettingsScreen';

type Props = NavigationData<'Settings'>;

const fakeProps: Props = {
  route: { key: '1234', name: 'Settings' },
  // @ts-ignore
  navigation: {},
};

describe('<SettingsScreen />', () => {
  it('matches the snapshot', () => {
    const tree = shallow(<SettingsScreen {...fakeProps} />);

    expect(tree).toMatchSnapshot();
  });
});
