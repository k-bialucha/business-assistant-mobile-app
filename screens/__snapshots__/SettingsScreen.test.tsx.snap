// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SettingsScreen /> matches the snapshot 1`] = `
<View>
  <Text>
    Settings
  </Text>
</View>
`;
